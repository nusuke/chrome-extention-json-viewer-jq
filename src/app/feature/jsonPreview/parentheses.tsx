type SurroundChars = { start: string; end: string };
/**
 * Object,Arrayを判別
 * @param value
 * @returns
 */
export function getSurroundCharactor(value: unknown): SurroundChars {
  if (Array.isArray(value)) {
    return { start: "[", end: "]" };
  } else if (typeof value === "object") {
    return { start: "{", end: "}" };
  } else {
    return { start: "", end: "" };
  }
}

/**
 * Array,Object型を{},[]で囲う
 * @param text
 * @param surroundChars
 * @returns
 */
export function surroundParentheses(
  text: JSX.Element[],
  surroundChars: SurroundChars
) {
  return (
    <>
      <span className="surroundChar--start">{surroundChars.start}</span>
      <span className="jsonObject">{text}</span>
      <span className="surroundChar--end">{surroundChars.end}</span>
    </>
  );
}
