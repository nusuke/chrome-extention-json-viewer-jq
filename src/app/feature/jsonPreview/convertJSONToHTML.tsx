import { logger } from "../../../lib/logger";
import { getSurroundCharactor, surroundParentheses } from "./parentheses";

/**
 * jsonを受け取り、key,value,中間オブジェクト毎にhtmlタグを付与
 * @param json
 * @returns
 */
export function convertJSONToHTML(json: JSON) {
  const nodes = [];

  for (const [key, value] of Object.entries(json)) {
    nodes.push(<span className="jsonKey">{key}</span>);
    if (typeof value === "object") {
      const surruondChar = getSurroundCharactor(value);
      nodes.push(
        <span className="jsonMiddleKey">
          {surroundParentheses(convertJSONToHTML(value), surruondChar)}
        </span>
      );
    } else {
      nodes.push(
        <span className="jsonValue">{convertJsonValueToHTML(value)}</span>
      );
    }
  }
  return nodes;
}

/**
 * jsonのValueを文字列、数値、URLで付与するclassを変える
 * @param value
 * @returns
 */
function convertJsonValueToHTML(value: unknown) {
  if (typeof value === "number") {
    return <span className="jsonValue--number">{value}</span>;
  } else if (typeof value === "string") {
    if (URL.canParse(value)) {
      return (
        <span className="jsonValue--url">
          <a href={value} target="_blank">
            {value}
          </a>
        </span>
      );
    } else {
      return <span className="jsonValue--string">{value}</span>;
    }
  } else {
    logger.debug("unexpected value:", value);
  }
}
