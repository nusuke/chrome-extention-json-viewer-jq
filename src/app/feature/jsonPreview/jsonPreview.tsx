import { useEffect } from "react";
import { convertJSONToHTML } from "./convertJSONToHTML";
import { getSurroundCharactor, surroundParentheses } from "./parentheses";
import { messageType } from "../../../background";

type P = { targetJson: JSON };
export const JsonPreview: React.FC<P> = (props: P) => {
  useEffect(() => {
    chrome.runtime.sendMessage({
      type: messageType.load,
      text: JSON.stringify(props.targetJson),
    });
  }, []);
  const surruondChar = getSurroundCharactor(props.targetJson);

  return (
    <div className="jsonPreview">
      {surroundParentheses(convertJSONToHTML(props.targetJson), surruondChar)}
    </div>
  );
};
