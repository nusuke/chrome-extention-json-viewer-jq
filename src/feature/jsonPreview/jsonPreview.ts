import { convertJSONToHTML } from "./convertJSONToHTML";
import { getSurroundCharactor, surroundParentheses } from "./parentheses";
import jq from "jq-web/jq.asm.bundle.js";

export const jsonPreview = (targetJson: JSON) => {
  console.log(targetJson);
  chrome.runtime.sendMessage({ type: "road", text: targetJson });

  const surruondChar = getSurroundCharactor(targetJson);
  // FIXME XSS
  document.body.innerHTML = surroundParentheses(
    convertJSONToHTML(targetJson),
    surruondChar
  );

  console.log("jq", targetJson);
  const jqQuery = ".contact";
  const res = jq.json(targetJson, jqQuery);
  console.log(res);
};
