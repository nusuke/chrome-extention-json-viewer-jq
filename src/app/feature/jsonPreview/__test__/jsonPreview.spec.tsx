import renderer from "react-test-renderer";
import { JsonPreview } from "../jsonPreview";

describe("jsonPreview snapshot", () => {
  global.URL.canParse = jest.fn();
  jest.spyOn(chrome.runtime, "sendMessage");
  it("json preview", async () => {
    const json = {
      posts: [{ id: 1, title: "json-server", author: "typicode" }],
      comments: [{ id: 1, body: "some comment", postId: 1 }],
      profile: {
        name: "John Doe",
        age: 30,
        url: "https://example.com",
        address: {
          city: "New York",
          postal_code: "10001",
          street: {
            name: "Broadway",
            number: 123,
            apartment: {
              floor: 5,
              number: 502,
            },
          },
        },
        contact: {
          email: "john.doe@example.com",
          phone: {
            home: "123-456-7890",
            work: "987-654-3210",
          },
        },
        interests: ["programming", "hiking", "reading"],
        test: [{ a: 1 }, { a: 2 }, { a: 3 }],
      },
    };

    const tree = renderer
      .create(<JsonPreview targetJson={JSON.parse(JSON.stringify(json))} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
