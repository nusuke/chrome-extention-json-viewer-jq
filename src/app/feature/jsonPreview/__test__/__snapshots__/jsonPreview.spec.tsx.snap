// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`jsonPreview snapshot json preview 1`] = `
<span>
  <span
    className="surroundChar--start"
  >
    {
  </span>
  <details
    open={true}
  >
    <summary />
    <span
      className="jsonObject"
    >
      <span
        className="jsonKey"
      >
        posts
      </span>
      <span
        className="jsonMiddleKey"
      >
        <span>
          <span
            className="surroundChar--start"
          >
            [
          </span>
          <details
            open={true}
          >
            <summary />
            <span
              className="jsonObject"
            >
              <span
                className="jsonKey"
              >
                0
              </span>
              <span
                className="jsonMiddleKey"
              >
                <span>
                  <span
                    className="surroundChar--start"
                  >
                    {
                  </span>
                  <details
                    open={true}
                  >
                    <summary />
                    <span
                      className="jsonObject"
                    >
                      <span
                        className="jsonKey"
                      >
                        id
                      </span>
                      <span
                        className="jsonValue"
                      >
                        <span
                          className="jsonValue--number"
                        >
                          1
                        </span>
                      </span>
                      <span
                        className="jsonKey"
                      >
                        title
                      </span>
                      <span
                        className="jsonValue"
                      >
                        <span
                          className="jsonValue--string"
                        >
                          json-server
                        </span>
                      </span>
                      <span
                        className="jsonKey"
                      >
                        author
                      </span>
                      <span
                        className="jsonValue"
                      >
                        <span
                          className="jsonValue--string"
                        >
                          typicode
                        </span>
                      </span>
                    </span>
                  </details>
                  <span
                    className="surroundChar--end"
                  >
                    }
                  </span>
                </span>
              </span>
            </span>
          </details>
          <span
            className="surroundChar--end"
          >
            ]
          </span>
        </span>
      </span>
      <span
        className="jsonKey"
      >
        comments
      </span>
      <span
        className="jsonMiddleKey"
      >
        <span>
          <span
            className="surroundChar--start"
          >
            [
          </span>
          <details
            open={true}
          >
            <summary />
            <span
              className="jsonObject"
            >
              <span
                className="jsonKey"
              >
                0
              </span>
              <span
                className="jsonMiddleKey"
              >
                <span>
                  <span
                    className="surroundChar--start"
                  >
                    {
                  </span>
                  <details
                    open={true}
                  >
                    <summary />
                    <span
                      className="jsonObject"
                    >
                      <span
                        className="jsonKey"
                      >
                        id
                      </span>
                      <span
                        className="jsonValue"
                      >
                        <span
                          className="jsonValue--number"
                        >
                          1
                        </span>
                      </span>
                      <span
                        className="jsonKey"
                      >
                        body
                      </span>
                      <span
                        className="jsonValue"
                      >
                        <span
                          className="jsonValue--string"
                        >
                          some comment
                        </span>
                      </span>
                      <span
                        className="jsonKey"
                      >
                        postId
                      </span>
                      <span
                        className="jsonValue"
                      >
                        <span
                          className="jsonValue--number"
                        >
                          1
                        </span>
                      </span>
                    </span>
                  </details>
                  <span
                    className="surroundChar--end"
                  >
                    }
                  </span>
                </span>
              </span>
            </span>
          </details>
          <span
            className="surroundChar--end"
          >
            ]
          </span>
        </span>
      </span>
      <span
        className="jsonKey"
      >
        profile
      </span>
      <span
        className="jsonMiddleKey"
      >
        <span>
          <span
            className="surroundChar--start"
          >
            {
          </span>
          <details
            open={true}
          >
            <summary />
            <span
              className="jsonObject"
            >
              <span
                className="jsonKey"
              >
                name
              </span>
              <span
                className="jsonValue"
              >
                <span
                  className="jsonValue--string"
                >
                  John Doe
                </span>
              </span>
              <span
                className="jsonKey"
              >
                age
              </span>
              <span
                className="jsonValue"
              >
                <span
                  className="jsonValue--number"
                >
                  30
                </span>
              </span>
              <span
                className="jsonKey"
              >
                url
              </span>
              <span
                className="jsonValue"
              >
                <span
                  className="jsonValue--string"
                >
                  https://example.com
                </span>
              </span>
              <span
                className="jsonKey"
              >
                address
              </span>
              <span
                className="jsonMiddleKey"
              >
                <span>
                  <span
                    className="surroundChar--start"
                  >
                    {
                  </span>
                  <details
                    open={true}
                  >
                    <summary />
                    <span
                      className="jsonObject"
                    >
                      <span
                        className="jsonKey"
                      >
                        city
                      </span>
                      <span
                        className="jsonValue"
                      >
                        <span
                          className="jsonValue--string"
                        >
                          New York
                        </span>
                      </span>
                      <span
                        className="jsonKey"
                      >
                        postal_code
                      </span>
                      <span
                        className="jsonValue"
                      >
                        <span
                          className="jsonValue--string"
                        >
                          10001
                        </span>
                      </span>
                      <span
                        className="jsonKey"
                      >
                        street
                      </span>
                      <span
                        className="jsonMiddleKey"
                      >
                        <span>
                          <span
                            className="surroundChar--start"
                          >
                            {
                          </span>
                          <details
                            open={true}
                          >
                            <summary />
                            <span
                              className="jsonObject"
                            >
                              <span
                                className="jsonKey"
                              >
                                name
                              </span>
                              <span
                                className="jsonValue"
                              >
                                <span
                                  className="jsonValue--string"
                                >
                                  Broadway
                                </span>
                              </span>
                              <span
                                className="jsonKey"
                              >
                                number
                              </span>
                              <span
                                className="jsonValue"
                              >
                                <span
                                  className="jsonValue--number"
                                >
                                  123
                                </span>
                              </span>
                              <span
                                className="jsonKey"
                              >
                                apartment
                              </span>
                              <span
                                className="jsonMiddleKey"
                              >
                                <span>
                                  <span
                                    className="surroundChar--start"
                                  >
                                    {
                                  </span>
                                  <details
                                    open={true}
                                  >
                                    <summary />
                                    <span
                                      className="jsonObject"
                                    >
                                      <span
                                        className="jsonKey"
                                      >
                                        floor
                                      </span>
                                      <span
                                        className="jsonValue"
                                      >
                                        <span
                                          className="jsonValue--number"
                                        >
                                          5
                                        </span>
                                      </span>
                                      <span
                                        className="jsonKey"
                                      >
                                        number
                                      </span>
                                      <span
                                        className="jsonValue"
                                      >
                                        <span
                                          className="jsonValue--number"
                                        >
                                          502
                                        </span>
                                      </span>
                                    </span>
                                  </details>
                                  <span
                                    className="surroundChar--end"
                                  >
                                    }
                                  </span>
                                </span>
                              </span>
                            </span>
                          </details>
                          <span
                            className="surroundChar--end"
                          >
                            }
                          </span>
                        </span>
                      </span>
                    </span>
                  </details>
                  <span
                    className="surroundChar--end"
                  >
                    }
                  </span>
                </span>
              </span>
              <span
                className="jsonKey"
              >
                contact
              </span>
              <span
                className="jsonMiddleKey"
              >
                <span>
                  <span
                    className="surroundChar--start"
                  >
                    {
                  </span>
                  <details
                    open={true}
                  >
                    <summary />
                    <span
                      className="jsonObject"
                    >
                      <span
                        className="jsonKey"
                      >
                        email
                      </span>
                      <span
                        className="jsonValue"
                      >
                        <span
                          className="jsonValue--string"
                        >
                          john.doe@example.com
                        </span>
                      </span>
                      <span
                        className="jsonKey"
                      >
                        phone
                      </span>
                      <span
                        className="jsonMiddleKey"
                      >
                        <span>
                          <span
                            className="surroundChar--start"
                          >
                            {
                          </span>
                          <details
                            open={true}
                          >
                            <summary />
                            <span
                              className="jsonObject"
                            >
                              <span
                                className="jsonKey"
                              >
                                home
                              </span>
                              <span
                                className="jsonValue"
                              >
                                <span
                                  className="jsonValue--string"
                                >
                                  123-456-7890
                                </span>
                              </span>
                              <span
                                className="jsonKey"
                              >
                                work
                              </span>
                              <span
                                className="jsonValue"
                              >
                                <span
                                  className="jsonValue--string"
                                >
                                  987-654-3210
                                </span>
                              </span>
                            </span>
                          </details>
                          <span
                            className="surroundChar--end"
                          >
                            }
                          </span>
                        </span>
                      </span>
                    </span>
                  </details>
                  <span
                    className="surroundChar--end"
                  >
                    }
                  </span>
                </span>
              </span>
              <span
                className="jsonKey"
              >
                interests
              </span>
              <span
                className="jsonMiddleKey"
              >
                <span>
                  <span
                    className="surroundChar--start"
                  >
                    [
                  </span>
                  <details
                    open={true}
                  >
                    <summary />
                    <span
                      className="jsonObject"
                    >
                      <span
                        className="jsonKey"
                      >
                        0
                      </span>
                      <span
                        className="jsonValue"
                      >
                        <span
                          className="jsonValue--string"
                        >
                          programming
                        </span>
                      </span>
                      <span
                        className="jsonKey"
                      >
                        1
                      </span>
                      <span
                        className="jsonValue"
                      >
                        <span
                          className="jsonValue--string"
                        >
                          hiking
                        </span>
                      </span>
                      <span
                        className="jsonKey"
                      >
                        2
                      </span>
                      <span
                        className="jsonValue"
                      >
                        <span
                          className="jsonValue--string"
                        >
                          reading
                        </span>
                      </span>
                    </span>
                  </details>
                  <span
                    className="surroundChar--end"
                  >
                    ]
                  </span>
                </span>
              </span>
              <span
                className="jsonKey"
              >
                test
              </span>
              <span
                className="jsonMiddleKey"
              >
                <span>
                  <span
                    className="surroundChar--start"
                  >
                    [
                  </span>
                  <details
                    open={true}
                  >
                    <summary />
                    <span
                      className="jsonObject"
                    >
                      <span
                        className="jsonKey"
                      >
                        0
                      </span>
                      <span
                        className="jsonMiddleKey"
                      >
                        <span>
                          <span
                            className="surroundChar--start"
                          >
                            {
                          </span>
                          <details
                            open={true}
                          >
                            <summary />
                            <span
                              className="jsonObject"
                            >
                              <span
                                className="jsonKey"
                              >
                                a
                              </span>
                              <span
                                className="jsonValue"
                              >
                                <span
                                  className="jsonValue--number"
                                >
                                  1
                                </span>
                              </span>
                            </span>
                          </details>
                          <span
                            className="surroundChar--end"
                          >
                            }
                          </span>
                        </span>
                      </span>
                      <span
                        className="jsonKey"
                      >
                        1
                      </span>
                      <span
                        className="jsonMiddleKey"
                      >
                        <span>
                          <span
                            className="surroundChar--start"
                          >
                            {
                          </span>
                          <details
                            open={true}
                          >
                            <summary />
                            <span
                              className="jsonObject"
                            >
                              <span
                                className="jsonKey"
                              >
                                a
                              </span>
                              <span
                                className="jsonValue"
                              >
                                <span
                                  className="jsonValue--number"
                                >
                                  2
                                </span>
                              </span>
                            </span>
                          </details>
                          <span
                            className="surroundChar--end"
                          >
                            }
                          </span>
                        </span>
                      </span>
                      <span
                        className="jsonKey"
                      >
                        2
                      </span>
                      <span
                        className="jsonMiddleKey"
                      >
                        <span>
                          <span
                            className="surroundChar--start"
                          >
                            {
                          </span>
                          <details
                            open={true}
                          >
                            <summary />
                            <span
                              className="jsonObject"
                            >
                              <span
                                className="jsonKey"
                              >
                                a
                              </span>
                              <span
                                className="jsonValue"
                              >
                                <span
                                  className="jsonValue--number"
                                >
                                  3
                                </span>
                              </span>
                            </span>
                          </details>
                          <span
                            className="surroundChar--end"
                          >
                            }
                          </span>
                        </span>
                      </span>
                    </span>
                  </details>
                  <span
                    className="surroundChar--end"
                  >
                    ]
                  </span>
                </span>
              </span>
            </span>
          </details>
          <span
            className="surroundChar--end"
          >
            }
          </span>
        </span>
      </span>
    </span>
  </details>
  <span
    className="surroundChar--end"
  >
    }
  </span>
</span>
`;
